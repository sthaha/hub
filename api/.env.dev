ENVIRONMENT="development"

POSTGRESQL_HOST="localhost"
POSTGRESQL_PORT="5432"
POSTGRESQL_DATABASE="hub"
POSTGRESQL_USER="postgres"
POSTGRESQL_PASSWORD="postgres"
