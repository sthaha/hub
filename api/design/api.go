package design

import (
	. "goa.design/goa/v3/dsl"
	cors "goa.design/plugins/v3/cors/dsl"
)

var _ = API("hub", func() {
	Title("Tekton Hub")
	Description("HTTP services for managing Tekton Hub")
	Version("1.0")
	Meta("swagger:example", "false")
	Server("hub", func() {
		Services("category", "swagger")
		Host("localhost", func() {
			URI("http://localhost:8000")
		})
	})
	cors.Origin("*", func() {
		cors.Headers("Content-Type")
		cors.Methods("GET")
	})
})
