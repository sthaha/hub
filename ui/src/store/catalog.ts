import { Instance, types } from 'mobx-state-tree';
import { IconName } from '../icons';

export const Catalog = types
  .model({
    id: types.identifierNumber,
    name: types.optional(types.string, ''),
    type: types.optional(types.string, ''),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    },
    icon() {
      if (self.type === 'official') return IconName.cat;
      else if (self.type === 'verified') return IconName.certificate;
      else if (self.type === 'community') return IconName.user;
    }
  }));

export type ICatalog = Instance<typeof Catalog>;
export type ICatalogStore = Instance<typeof CatalogStore>;

export const CatalogStore = types
  .model({
    items: types.map(Catalog)
  })

  .actions((self) => ({
    add(item: ICatalog) {
      self.items.put({ id: item.id, name: item.name, type: item.type });
    },
    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    }
  }))

  .views((self) => ({
    get values() {
      return Array.from(self.items.values());
    },

    get selected() {
      const list = new Set();
      self.items.forEach((c: ICatalog) => {
        if (c.selected) {
          list.add(c.id);
        }
      });

      return list;
    }
  }));
