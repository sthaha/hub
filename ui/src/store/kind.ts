import { types, Instance } from 'mobx-state-tree';

export const Kind = types
  .model({
    name: types.identifier,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type IKind = Instance<typeof Kind>;
export type IKindStore = Instance<typeof KindStore>;

export const KindStore = types
  .model({
    kinds: types.map(Kind)
  })

  .actions((self) => ({
    add(item: string) {
      self.kinds.put({ name: item, selected: false });
    },

    clearSelected() {
      self.kinds.forEach((k) => {
        k.selected = false;
      });
    }
  }))

  .views((self) => ({
    get items() {
      return Array.from(self.kinds.values());
    },

    get selected() {
      const list = new Set();
      self.kinds.forEach((c: IKind) => {
        if (c.selected) {
          list.add(c.name);
        }
      });

      return list;
    }
  }));
