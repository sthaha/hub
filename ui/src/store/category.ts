import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Tag = types.model('Tags', {
  id: types.identifierNumber,
  name: types.string
});

export const Category = types
  .model('Category', {
    id: types.identifierNumber,
    name: types.string,
    tags: types.array(types.reference(Tag)),
    selected: false
  })

  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ITag = Instance<typeof Tag>;
export type ICategory = Instance<typeof Category>;
export type ICategoryStore = Instance<typeof CategoryStore>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.map(Category),
    tags: types.optional(types.map(Tag), {}),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.size;
    },

    get values() {
      return Array.from(self.items.values());
    },

    get selectedTags() {
      return new Set(
        Array.from(self.items.values())
          .filter((c: ICategory) => c.selected)
          .reduce((acc: number[], c: ICategory) => [...acc, ...c.tags.map((t: ITag) => t.id)], [])
      );
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.put(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;

        const json = yield api.categories();

        // adding the tags to the store - normalized
        const tags: ITag[] = json.data.flatMap((item: ICategory) => item.tags);

        tags.forEach((t) => self.tags.put(t));

        // creating the model only after the store has the tags normalized
        const categories: ICategory[] = json.data.map((c: ICategory) => ({
          id: c.id,
          name: c.name,
          tags: c.tags.map((tag: ITag) => tag.id)
        }));

        categories.forEach((c: ICategory) => self.add(c));
      } catch (err) {
        self.err = err.toString();
      }

      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));
