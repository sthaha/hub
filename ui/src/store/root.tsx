import React, { ReactChild, ReactChildren } from 'react';
import { types, getEnv, Instance } from 'mobx-state-tree';
import { CategoryStore } from './category';
import { ResourceStore } from './resource';
import { Hub, Api } from '../api';

export const Root = types.model('Root', {}).views((self) => ({
  get api(): Api {
    return getEnv(self).api;
  },
  get categories() {
    return getEnv(self).categories;
  },
  get resources() {
    return getEnv(self).resources;
  }
}));

export type IRoot = Instance<typeof Root>;

let RootContext: React.Context<IRoot>;
export const useMst = () => React.useContext(RootContext);

export const initRootStore = (api: Api) => {
  const categories = CategoryStore.create({}, { api });
  const resources = ResourceStore.create({ catalogs: {}, kinds: {} }, { api });
  return Root.create({}, { api, categories, resources });
};

interface Props {
  children: ReactChild | ReactChildren;
}

export const createProviderAndStore = (api?: Api) => {
  const root = initRootStore(api || new Hub());
  RootContext = React.createContext(root);

  const Provider = ({ children }: Props) => (
    <RootContext.Provider value={root}> {children} </RootContext.Provider>
  );
  return { Provider, root };
};

export const createProvider = (api?: Api) => {
  const { Provider } = createProviderAndStore(api);
  return Provider;
};
