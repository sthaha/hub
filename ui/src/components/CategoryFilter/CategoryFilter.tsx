import React from 'react';
import { useObserver } from 'mobx-react';
import Filter from '../Filter/Filter';
import { useMst } from '../../store/root';

const CategoryFilter: React.FC = () => {
  const { categories } = useMst();
  return useObserver(() => <Filter store={categories} header="Categories" />);
};

export default CategoryFilter;
