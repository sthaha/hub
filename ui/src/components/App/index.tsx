import React from 'react';
import { observer } from 'mobx-react';
import '@patternfly/react-core/dist/styles/base.css';

import CategoryFilter from '../CategoryFilter/CategoryFilter';
import { createProvider } from '../../store/root';

const Provider = createProvider();

const App: React.FC = observer(() => {
  return (
    <Provider>
      <div className="App">
        <CategoryFilter />
      </div>
    </Provider>
  );
});

export default App;
