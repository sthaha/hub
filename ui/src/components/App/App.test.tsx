import React from 'react';
import App from '.';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { FakeHub } from '../../api/testutil';
import CategoryFilter from '../CategoryFilter/CategoryFilter';
import { createProviderAndStore } from '../../store/root';

const TESTDATA_DIR = `src/store/testdata`;
const api = new FakeHub(TESTDATA_DIR);
const { Provider } = createProviderAndStore(api);

describe('App', () => {
  it('should render the component correctly and match the snapshot', (done) => {
    const app = renderer.create(
      <Provider>
        <div className="App">
          <CategoryFilter />
        </div>
      </Provider>
    );

    expect(app.toJSON()).toMatchSnapshot();
    done();
  });

  it('should find the categoryFilter component and match the count', () => {
    const component = shallow(<App />);
    expect(component.find(CategoryFilter).length).toEqual(1);
  });
});
