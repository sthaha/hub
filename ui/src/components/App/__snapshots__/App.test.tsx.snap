// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the component correctly and match the snapshot 1`] = `
Array [
  " ",
  <div
    className="App"
  >
    <div
      className="hub-filter"
    >
      <div
        className="pf-l-grid pf-m-all-6-col-on-sm pf-m-all-4-col-on-md pf-m-all-3-col-on-lg pf-m-all-1-col-on-2xl"
      >
        <div
          className="pf-l-grid__item pf-m-1-col pf-m-1-row hub-filter-header"
        >
          <h1
            className=""
            data-pf-content={true}
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Categories
          </h1>
        </div>
        <div
          className="pf-l-grid__item pf-m-2-row"
        >
          <button
            aria-disabled={false}
            aria-label="Clear"
            className="pf-c-button pf-m-plain"
            data-ouia-component-id={0}
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                transform=""
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        className="pf-l-grid"
      >
        <div
          className="pf-l-grid__item hub-filter-checkboxes"
        />
      </div>
    </div>
  </div>,
  " ",
]
`;
